libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -36,18 +36,66 @@
 #include <libsolutil/CommonData.h>
 #include <libsolutil/Whiskers.h>
 #include <libsolutil/StringUtils.h>
+#include <libsolutil/Algorithms.h>
 
 #include <liblangutil/SourceReferenceFormatter.h>
 
 #include <boost/range/adaptor/map.hpp>
 
 #include <sstream>
+#include <variant>
 
 using namespace std;
 using namespace solidity;
 using namespace solidity::util;
 using namespace solidity::frontend;
 
+namespace
+{
+
+void verifyCallGraph(
+	set<CallableDeclaration const*, ASTNode::CompareByID> const& _expectedCallables,
+	set<FunctionDefinition const*> _generatedFunctions
+)
+{
+	for (auto const& expectedCallable: _expectedCallables)
+		if (auto const* expectedFunction = dynamic_cast<FunctionDefinition const*>(expectedCallable))
+		{
+			solAssert(
+				_generatedFunctions.count(expectedFunction) == 1 || expectedFunction->isConstructor(),
+				"No code generated for function " + expectedFunction->name() + "even though it is not a constructor."
+			);
+			_generatedFunctions.erase(expectedFunction);
+		}
+
+	solAssert(
+		_generatedFunctions.size() == 0,
+		"Of the generated functions " + toString(_generatedFunctions.size()) + " are not in the call graph."
+	);
+}
+
+void collectReachableCallables(
+	FunctionCallGraphBuilder::ContractCallGraph const& _graph,
+	FunctionCallGraphBuilder::Node _root,
+	set<CallableDeclaration const*, ASTNode::CompareByID>& o_reachableCallables
+)
+{
+	using Node = FunctionCallGraphBuilder::Node;
+
+	set<Node const*> reachableNodes = BreadthFirstSearch<Node const*>{{&_root}}.run(
+		[&_graph](Node const* _node, auto&& _addChild) {
+			for (Node const& _child: valueOrDefault(_graph.edges, *_node, {}))
+				_addChild(&_child);
+		}
+	).visited;
```

Comment:
The graph by definition is connected, right? So I think it suffices to return the keys.

Reply:
Well, it could be a disconnected graph. But it depends more on how we plan to use it than on the definition so if you say it's fine right now, I'm leaving it at that :)

Reply:
How can it be disconnected? If it is disconnected, then it is a bug.

Reply:
I meant that the mathematical graph definition includes both connected and disconnected graphs.

Also, our function call graph could easily contain isolated subgraphs if we wanted it to - e.g. if we made it visit all the internal functions in a contract. Whether it's a bug depends only on whether this is what we want it to do or not.